specification/ws.extensions/x-kaazing-ping-pong/SPEC.md

```diff
@@ -0,0 +1,73 @@
+
+### Requirements
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Table of Contents
+
+  * [Introduction](#introduction)
+  * [Handshake Negotiation](#handshake-negotiation)
+  * [Protocol Definitions: extended PING and PONG](#ex-ping-pong)
+  * [References](#references)
+
+### Introduction
+
+This extension applies only when the WebSocket connection is successfully established with the “x-kaazing-ping-pong” extension being affirmatively negotiated. 
+
+The intention of the extension is to allow the Kaazing WebSocket javascript client library to send and receive PING and PONG control frames over native (Browser implemented) WebSockets, which do not expose the standard WebSocket PING and PONG frames to the consumer.
+
+### Handshake Negotiation
+
+The “x-kaazing-ping-pong” WebSocket extension is negotiated during the initial WebSocket handshake.
+
+
+```
+	GET /transport HTTP/1.1\r\n
+	Upgrade: websocket\r\n
+	…
+	Sec-WebSocket-Extension: x-kaazing-ping-pong
+	…
+```
+
+The extension name is specified on the initial connection handshake. 
+
+```
+	HTTP/1.1 101 Switching Protocols\r\n
+	Upgrade: websocket\r\n
+	…
+	Sec-WebSocket-Extension:x-kaazing-ping-pong;<control_bytes>
+	…
+```
+<`control_bytes`> designates four bytes represented in hexadecimal notation, as described in the WebSocket Extension Control Frame Injection specification.
+
+Once the extension is negotiated successfully, the server MAY send extended PING and PONG control frames (as defined below) rather than the standard PING and PONG frames defined by the WebSocket Protocol [WSP], in order to make them visible to the Kaazing WebSocket client library. The server MUST respond to an extended PING frame with an extended PONG frame. But it MUST still respond to standard WebSocket PING frame with a standard PONG. The client MUST respond to an extended PING control frame with an extended PONG frame. The server may also assume that the client will respond to a standard WebSocket PING frame with a standard PONG (since the Browser native WebSocket implementation will take care of this).
+
+
+### Protocol Definitions: extended PING and PONG
+
+Once the extension is negotiated, all WebSocket frames are delivered unmodified, but WebSocket frames may include extended PING and PONG frames. These are websocket text frames distinguished by a server-negotiated control frame control byte sequence at the start of the payload, as defined by the WebSocket Extension Control Frame Injection specification. The format of the payload of these frames is the control bytes followed by a standard RFC 6455 PING or PONG frame with the fin bit unset (so it's a valid UTF) and no application data (zero data length):
```

Comment:
We'll need to add a link to the Control Frame Extension specification. Unfortunately that document is currently on Google docs (at https://docs.google.com/document/d/1ujCKQeKP-abibtDtiHchMW0SKLYfrk9QGl-ccginxMM/edit#heading=h.2eetvmjc7sjt) which is not a public location. Please add it to your work as SPEC.md under directory specifications/ws.extensions/control-frame-injection (this directory will be a sibling of the x-kaazing-ping-pong directory). Then you can add a link to it in this document. Sorry about this extra work, but at least you got out of having to do WSE tests :)